firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -235,50 +230,81 @@ private void persistFid(String fid) throws FirebaseInstallationsException {
+      throws FirebaseInstallationsException {
+    Boolean isNetworkCallRequired = false;
+
+    synchronized (persistedFid) {
+      PersistedFidEntry persistedFidEntry = persistedFid.readPersistedFidEntryValue();
+
+      if (persistedFidEntry.isNotGenerated()) {
+        throw new FirebaseInstallationsException(
+            "Local storage has no persisted Fid entry.",
+            FirebaseInstallationsException.Status.CLIENT_ERROR);
+      }
+
+      // If FID registration status is REGISTERED and auth token is expired, or registration status
+      // is UNREGISTERED network call iks required.
+      if ((persistedFidEntry.isRegistered() && isAuthTokenExpired(persistedFidEntry))
+          || persistedFidEntry.isUnregistered()) {
+        isNetworkCallRequired = true;
+
+        // Update FID registration status to PENDING to avoid multiple network calls to FIS Servers.
+        persistedFid.insertOrUpdatePersistedFidEntry(
+            persistedFidEntry
+                .toBuilder()
+                .setRegistrationStatus(RegistrationStatus.PENDING)
+                .build());
+      }
     }
 
-    return Tasks.forResult(fid);
-  }
+    PersistedFidEntry updatedPersistedFidEntry = persistedFid.readPersistedFidEntryValue();
 
-  private void updateAwaitListenerIfRegisteredFid(
-      PersistedFidEntry persistedFidEntry, AwaitListener listener) {
-    if (listener != null
-        && persistedFidEntry.getRegistrationStatus() == RegistrationStatus.REGISTERED) {
-      listener.onSuccess();
+    if (updatedPersistedFidEntry.isNotGenerated()) {
+      throw new FirebaseInstallationsException(
+          "Local storage has no persisted Fid entry.",
+          FirebaseInstallationsException.Status.CLIENT_ERROR);
     }
-  }
 
-  /**
-   * Registers the FID with FIS servers in a background thread and updates the listener on
-   * completion.
-   */
-  private void executeFidRegistration(PersistedFidEntry persistedFidEntry, AwaitListener listener) {
-    Task<Void> task = Tasks.call(executor, () -> registerAndSaveFid(persistedFidEntry));
-    if (listener != null) {
-      task.addOnCompleteListener(listener);
+    String fid = updatedPersistedFidEntry.getFirebaseInstallationId();
+
+    if (!isNetworkCallRequired) {
+
+      // If the Fid is registered, update the listener if awaiting
+      if (listener != null && updatedPersistedFidEntry.isRegistered()) {
+        listener.onSuccess();
+      }
+      return Tasks.forResult(fid);
     }
-  }
 
-  private void updatePersistedFidWithPendingStatus(String fid) {
-    persistedFid.insertOrUpdatePersistedFidEntry(
-        PersistedFidEntry.builder()
-            .setFirebaseInstallationId(fid)
-            .setRegistrationStatus(RegistrationStatus.PENDING)
-            .build());
+    // If fid registration for this firebase installation is incomplete without an expired auth
+    // token, execute Fid registration.
+    if (!isAuthTokenExpired(updatedPersistedFidEntry)) {
```

Comment:
isn't this being checked twice in the same body?

Reply:
Yes, it is. First time its checked to evaluate if we need network calls. Second check helps to decide what type of network call is required.

Network call should be made : If UNREGISTERED or AuthToken Expired
Before triggering the call, we update the status to PENDING.

So, we rely on  AuthToken Expired check. 

If AuthToken not Expired: REGISTER Fid
else: Refresh Auth Token

Reply:
I'm having a really really hard time understanding the flow of this method. I don't know if it's just that I lack the domain knowledge or familiarity with the terms, but the multiple return points, double checking in and out of the synch block, and sometimes-async kickoffs are breaking my brain.